### PR TITLE
Spec style

### PR DESCRIPTION
**Pull request by mhevery**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

### CODE REVIEW

File: packages/core/src/render3/index.ts [hunk 1]
```diff
@@ -67,6 +69,7 @@ export {
   embeddedViewEnd as v,
   detectChanges,
   markDirty,
+  tick
```
Comment (vicb) [inline packages/core/src/render3/index.ts]: ?

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -845,6 +845,29 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
   }
 }
 
+/**
+ * Set class on an element.
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: class ? element ?

File: packages/core/src/render3/instructions.ts [hunk 3]
```diff
@@ -831,7 +831,7 @@ function generatePropertyAliases(lNodeFlags: number, direction: BindingDirection
  *        renaming as part of minification.
  * @param value A value indicating if a given class should be added or removed.
  */
-export function elementClass<T>(index: number, className: string, value: T | NO_CHANGE): void {
+export function elementClassNamed<T>(index: number, className: string, value: T | NO_CHANGE): void {
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: the `elementClass` variant seems to be the most specific / counter-intuitive. Keep `elementClass()` for this one and find a meaningful name for the other one ?

File: packages/core/src/render3/instructions.ts [hunk 4]
```diff
@@ -845,6 +845,29 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
   }
 }
 
+/**
+ * Set class on an element.
+ *
+ * This instruction is meant to handle the [class]="exp" usage.
+ *
+ * In its simplest form the `elementClass` instruction simply writes the value
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: "In its simplest form" -> what does this means ?

File: packages/core/src/render3/instructions.ts [hunk 5]
```diff
@@ -845,6 +845,29 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+ * to the element`s `className` property. However we don't want to use the
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: use single quote for "element's"

File: packages/core/src/render3/instructions.ts [hunk 6]
```diff
@@ -845,6 +845,29 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+ * `elementProperty` instruction because `elementClass` should understand
+ * merging classes as in `<div class="foo" [class]="exp">`.
+ *
+ * @param index The index of the element to update in the data array
+ * @param value A value indicating if a given class should be added or removed.
+ */
+export function elementClass<T>(index: number, value: T | NO_CHANGE): void {
+  if (value !== NO_CHANGE) {
+    // This is a naive implementation which simple writes value to the `className`. In the future
+    // we will add logic here which would work with the animation code.
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: either "TODO" or remove ?

File: packages/core/src/render3/instructions.ts [hunk 7]
```diff
@@ -880,6 +903,38 @@ export function elementStyle<T>(
   }
 }
 
+/**
+ * Set style on an element.
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: Too vague ?

File: packages/core/src/render3/instructions.ts [hunk 8]
```diff
@@ -845,6 +845,29 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+/**
+ * Set class on an element.
+ *
+ * This instruction is meant to handle the [class]="exp" usage.
+ *
+ * In its simplest form the `elementClass` instruction simply writes the value
+ * to the element`s `className` property. However we don't want to use the
+ * `elementProperty` instruction because `elementClass` should understand
+ * merging classes as in `<div class="foo" [class]="exp">`.
+ *
```
Comment (kara) [inline packages/core/src/render3/instructions.ts]: This comment implies that in the above example, the output should have the `foo` class and whichever classes are returned by `exp`. However, the below implementation would overwrite the `foo` class, like Angular does currently. Is the plan to overwrite the class name to match current behavior for now, then to change the implementation to merge the classes down the line? Or will the compiler switch from generating `elementProperty` to generating `elementClass`, which would imply the implementation below should change?  I think the comment could use some clarification.

File: packages/core/test/render3/compiler_canonical/elements_spec.ts [hunk 9]
```diff
@@ -264,5 +264,36 @@ describe('elements', () => {
       $r3$.ɵdetectChanges(comp);
       expect(toHtml(comp)).toEqual('<div class="" id="changed1" style="color: red;"></div>');
     });
+
+    it('should bind [class] and [style]', () => {
```
Comment (vicb) [inline packages/core/test/render3/compiler_canonical/elements_spec.ts]: bind to ... ?

File: packages/core/test/render3/instructions_spec.ts [hunk 10]
```diff
@@ -47,19 +47,41 @@ describe('instructions', () => {
     });
   });
 
-  describe('elementStyle', () => {
+  describe('elementStyleNamed', () => {
```
Comment (vicb) [inline packages/core/test/render3/instructions_spec.ts]: Use words instead of function names here and below and for class ?

File: packages/core/test/render3/render_util.ts [hunk 11]
```diff
@@ -59,15 +74,44 @@ export class TemplateFixture {
...
+  update(): void {
+    tick(this.component);
+    this.requestAnimationFrame.flush();
   }
 }
 
+///////////////////////////////////////////////////////////////////////////////////
+// The methods bellow use global state and we should stop using them.
+// Fixtures above are preferred way of testing Components and Templates
+///////////////////////////////////////////////////////////////////////////////////
```
Comment (vicb) [inline packages/core/test/render3/render_util.ts]: Use a `@deprecated` on the function instead. I guess nobody will ever see this comment ?

File: packages/core/test/render3/render_util.ts [hunk 12]
```diff
@@ -59,15 +74,44 @@ export class TemplateFixture {
         this.hostNode.native, updateBlock || this.updateBlock, null !, domRendererFactory3,
         this.hostNode);
   }
+}
+
+
+/**
+ * Fixture for testing Components in a convenient way.
+ */
+export class ComponentFixture<T> extends BaseFixture {
```
Comment (vicb) [inline packages/core/test/render3/render_util.ts]: Should be a different commit ?

File: packages/core/src/render3/index.ts (hunk 1 above)
Comment (vicb) [reply packages/core/src/render3/index.ts]: preserve the trailing coma ?

File: packages/core/src/render3/instructions.ts [hunk 13]
```diff
@@ -856,6 +856,26 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
   }
 }
 
+/**
+ * Set class on an element.
+ *
+ * This instruction is meant to handle the [class]="exp" usage.
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: back quote around code "`[class]="exp"`"

File: packages/core/src/render3/instructions.ts [hunk 14]
```diff
@@ -856,6 +856,26 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+ *
+ * `elementClass` instruction writes the value to the "element`s" `className` property.
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: element\`s should be a single quote

File: packages/core/src/render3/instructions.ts (hunk 2 above)
Comment (vicb) [reply packages/core/src/render3/instructions.ts]: "Set the className property on a DOM element" ?

File: packages/core/src/render3/instructions.ts [hunk 15]
```diff
@@ -856,6 +856,26 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+ *
+ * @param index The index of the element to update in the data array
+ * @param value A value indicating if a given class should be added or removed.
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: "A value indicating if a given class should be added or removed."
Describe the shape of `value` ans the meaning of keys & values ?

File: packages/core/src/render3/instructions.ts [hunk 16]
```diff
@@ -856,6 +856,26 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+ */
+export function elementClass<T>(index: number, value: T | NO_CHANGE): void {
+  if (value !== NO_CHANGE) {
+    // TODO: This is a naive implementation which simply writes value to the `className`. In the future
+    // we will add logic here which would work with the animation code.
+    const lElement = data[index] as LElementNode;
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: nit: I usually prefer the `const lElement: LElementNode = data[index];` form

Comment (vicb) [reply packages/core/src/render3/instructions.ts]: maybe get the native element in a const ?

File: packages/core/src/render3/instructions.ts [hunk 17]
```diff
@@ -856,6 +856,26 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+    isProceduralRenderer(renderer) ? renderer.setProperty(lElement.native, 'className', value) :
+                                     lElement.native.className = stringify(value);
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: should it be `['className']` here ?

File: packages/core/src/render3/instructions.ts (hunk 7 above)
Comment (vicb) [reply packages/core/src/render3/instructions.ts]: See the comment I added for class above

File: packages/core/src/render3/instructions.ts [hunk 18]
```diff
@@ -891,6 +911,33 @@ export function elementStyle<T>(
...
 }
 
+/**
+ * Set style on an element.
+ *
+ * This instruction is meant to handle the [style]="exp" usage.
+ *
+ *
+ * @param index The index of the element to update in the data array
+ * @param value A value indicating if a given class should be added or removed.
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: class -> style. "added or removed" is wrong here

Comment (vicb) [reply packages/core/src/render3/instructions.ts]: plus see my comment about `value` on class above

File: packages/core/src/render3/instructions.ts [hunk 19]
```diff
@@ -891,6 +911,33 @@ export function elementStyle<T>(
...
+ */
+export function elementStyle<T>(index: number, value: T | NO_CHANGE): void {
+  if (value !== NO_CHANGE) {
+    // TODO: This is a naive implementation which simply writes value to the `style`. In the future
+    // we will add logic here which would work with the animation code.
+    const lElement = data[index] as LElementNode;
+    if (isProceduralRenderer(renderer)) {
+      renderer.setProperty(lElement.native, 'style', value);
+    } else {
+      const style = lElement.native.style;
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: should probably be `['style']` here. Also everywhere else in the code. 
Handle this as a separate fix commit ?

File: packages/core/src/render3/instructions.ts [hunk 20]
```diff
@@ -891,6 +911,33 @@ export function elementStyle<T>(
...
+ *
+ *
+ * @param index The index of the element to update in the data array
+ * @param value A value indicating if a given class should be added or removed.
+ */
+export function elementStyle<T>(index: number, value: T | NO_CHANGE): void {
+  if (value !== NO_CHANGE) {
+    // TODO: This is a naive implementation which simply writes value to the `style`. In the future
+    // we will add logic here which would work with the animation code.
+    const lElement = data[index] as LElementNode;
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: same comments as above for the form + extracting native

File: packages/core/src/render3/instructions.ts [hunk 21]
```diff
@@ -891,6 +911,33 @@ export function elementStyle<T>(
...
+    if (isProceduralRenderer(renderer)) {
+      renderer.setProperty(lElement.native, 'style', value);
+    } else {
+      const style = lElement.native.style;
+      for (let styleName in value as any) {
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: I never remember which of `for...in` or `Object.keys` is more performant. Could you add a note to perf.md ?

File: packages/core/src/render3/instructions.ts [hunk 22]
```diff
@@ -891,6 +911,33 @@ export function elementStyle<T>(
...
+        const styleValue = (value as any)[styleName];
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: `const styleValue: any = ...`; ?

File: packages/core/src/render3/instructions.ts [hunk 23]
```diff
@@ -891,6 +911,33 @@ export function elementStyle<T>(
...
+        styleValue ? style.setProperty(styleName, styleValue) : style.removeProperty(styleName);
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: I would go for `styleValue == null` here.

File: packages/core/test/render3/compiler_canonical/elements_spec.ts [hunk 24]
```diff
@@ -264,5 +264,36 @@ describe('elements', () => {
...
       expect(toHtml(comp)).toEqual('<div class="" id="changed1" style="color: red;"></div>');
     });
+
+    it('should bind [class] and [style] to the element', () => {
+      type $StyleComponent$ = StyleComponent;
+
+      @Component(
+          {selector: 'style-comp', template: `<div [class]="classExp" [style]="styleExp"></div>`})
+      class StyleComponent {
+        classExp: string[]|string = 'some-name';
```
Comment (vicb) [inline packages/core/test/render3/compiler_canonical/elements_spec.ts]: Do we plan to support `{[name: string]: boolean}` ?

File: packages/core/test/render3/instructions_spec.ts (hunk 10 above)
Comment (vicb) [reply packages/core/test/render3/instructions_spec.ts]: 1) Change the rest of the file ?
2) "Setting styles on elements" ?

File: packages/core/test/render3/instructions_spec.ts [hunk 25]
```diff
@@ -55,19 +55,41 @@ describe('instructions', () => {
...
+      fixture.update(() => elementStyle(0, {'height': '10px'}));
+      expect(fixture.html).toEqual('<div style="background-color: red; height: 10px;"></div>');
+    });
+  });
+
+  describe('elementClass', () => {
+    const fixture = new TemplateFixture(createDiv);
+
+    it('should add class', () => {
+      fixture.update(() => elementClass(0, "new"));
```
Comment (vicb) [inline packages/core/test/render3/instructions_spec.ts]: test `(0, "multiple classes")` ?

File: packages/core/test/render3/render_util.ts [hunk 26]
```diff
@@ -104,6 +148,9 @@ export function resetDOM() {
   // TODO: assert that the global state is clean (e.g. ngData, previousOrParentNode, etc)
 }
 
+/**
+ * @deprecated use `TemplateFixture` or `ComponentFixture`
```
Comment (vicb) [inline packages/core/test/render3/render_util.ts]: :+1:

File: packages/core/src/render3/instructions.ts [hunk 27]
```diff
@@ -856,6 +856,26 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+ * Set class on an element.
+ *
+ * This instruction is meant to handle the [class]="exp" usage.
+ *
+ * `elementClass` instruction writes the value to the "element`s" `className` property.
+ *
+ * @param index The index of the element to update in the data array
+ * @param value A value indicating if a given class should be added or removed.
+ */
+export function elementClass<T>(index: number, value: T | NO_CHANGE): void {
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: does `<T>` makes sense here ? (same question for style)

**Review by kara — APPROVED** [review]
LGTM

File: packages/core/src/render3/index.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/render3/index.ts]: fixed

File: packages/core/src/render3/instructions.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: You are right, that is better. changed.

File: packages/core/src/render3/instructions.ts (hunk 7 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed to be consistent.

File: packages/core/test/render3/instructions_spec.ts (hunk 10 above)
Comment (mhevery) [reply packages/core/test/render3/instructions_spec.ts]: 1. happy to change the rest of the file.
2. Not sure Setting styles on elements is better. Currently it is clear I am testing a specific instruction, if I change it should I merge all instructions into single `describe`? That does not sound right.

File: packages/core/src/render3/instructions.ts (hunk 13 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed

File: packages/core/src/render3/instructions.ts (hunk 14 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: Originally it was without quotes. I added them, now you want single? but that would collide with apostrophe `'`.

File: packages/core/src/render3/instructions.ts (hunk 15 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: updated.

File: packages/core/src/render3/instructions.ts (hunk 16 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: done.

File: packages/core/src/render3/instructions.ts (hunk 17 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: I don't think it matters since closure compiler knows that `className` could potentially be on `Element` and so it does not rename, but I have changed it anyways.

File: packages/core/src/render3/instructions.ts (hunk 18 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed

File: packages/core/src/render3/instructions.ts (hunk 19 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed

Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed

File: packages/core/src/render3/instructions.ts (hunk 20 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed

File: packages/core/src/render3/instructions.ts (hunk 21 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed

File: packages/core/src/render3/instructions.ts (hunk 22 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed

File: packages/core/src/render3/instructions.ts (hunk 23 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed

File: packages/core/test/render3/compiler_canonical/elements_spec.ts (hunk 24 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/elements_spec.ts]: maybe, not sure yet...

File: packages/core/test/render3/instructions_spec.ts (hunk 25 above)
Comment (mhevery) [reply packages/core/test/render3/instructions_spec.ts]: fixed

File: packages/core/src/render3/PERF_NOTES.md [hunk 28]
```diff
@@ -50,4 +50,15 @@ export function getExported() { return exported; }
...
+Also writing to a property of `exports` might change its hidden class resulting in megamorphic access.
+
+## Iterating over Keys of an Object.
+
+https://jsperf.com/object-keys-vs-for-in-with-closure/3 implies that `Object.keys` is the fastest way of iterating 
+over properties of an object.
+
+```
+for (var i = 0, keys = Object.key(obj); i < keys.length; i++) {
+  cons key = keys[i];
```
Comment (vicb) [inline packages/core/src/render3/PERF_NOTES.md]: typo `const`.
:+1: for improving this file

File: packages/core/src/render3/PERF_NOTES.md [hunk 29]
```diff
@@ -50,4 +50,15 @@ export function getExported() { return exported; }
...
\ No newline at end of file
+Also writing to a property of `exports` might change its hidden class resulting in megamorphic access.
+
+## Iterating over Keys of an Object.
+
+https://jsperf.com/object-keys-vs-for-in-with-closure/3 implies that `Object.keys` is the fastest way of iterating 
+over properties of an object.
+
+```
+for (var i = 0, keys = Object.key(obj); i < keys.length; i++) {
```
Comment (vicb) [inline packages/core/src/render3/PERF_NOTES.md]: `.keys`

File: packages/core/src/render3/PERF_NOTES.md (hunk 28 above)
Comment (mhevery) [reply packages/core/src/render3/PERF_NOTES.md]: fixed

File: packages/core/src/render3/PERF_NOTES.md (hunk 29 above)
Comment (mhevery) [reply packages/core/src/render3/PERF_NOTES.md]: fixed

File: packages/core/src/render3/instructions.ts [hunk 30]
```diff
@@ -867,31 +890,65 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+    // TODO: This is a naive implementation which simply writes value to the `style`. In the future
+    // we will add logic here which would work with the animation code.
+    const lElement = data[index] as LElementNode;
+    if (isProceduralRenderer(renderer)) {
+      renderer.setProperty(lElement.native, 'style', value);
+    } else {
+      const style = lElement.native['style'];
+      for (let i = 0, keys = Object.keys(value); i < keys.length; i++) {
+        const styleName: string = keys[i];
+        const styleValue: any = (value as any)[styleName];
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: drop `( as any)` ?

File: packages/core/src/render3/instructions.ts [hunk 31]
```diff
@@ -867,31 +890,65 @@ export function elementClass<T>(index: number, className: string, value: T | NO_
...
+ *
+ *
+ * @param index The index of the element to update in the data array
+ * @param value A value indicating if a given style should be added or removed.
+ *   The expected shape of `value` is an object where keys are style names and the values
+ *   are their corresponding values to set. If value is falsy than the style is remove. An absence
+ *   of style does not cause that style to be removed. `NO_CHANGE` implies that no update should be
+ *   performed.
+ */
+export function elementStyle<T>(
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: remove `<T>` ?

Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: Per comment above I think we should keep the <T>.

Comment (vicb) [reply packages/core/src/render3/instructions.ts]: Does it apply with `{[styleName: string]: any} | NO_CHANGE` ?

File: packages/core/src/render3/instructions.ts (hunk 30 above)
Comment (vicb) [reply packages/core/src/render3/instructions.ts]: Still true with `{[styleName: string]: any} | NO_CHANGE` ?

**Review by vicb — APPROVED** [review]
Thanks for the update. Added a few more comments

**Comment by mhevery** [comment]
http://test/OCL:188976127:BASE:188976132:1521000793900:5b07f027
